### PR TITLE
fix: check for duplicate tokens in the balance proofs

### PR DESCRIPTION
**Pull request by wborgeaud**
@invocamanman found a potential vulnerability in the PP when `multi_batch_header.balances_proofs` contains duplicate tokens. This PR fixes this by ensuring there are no duplicates.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

**Review by invocamanman — COMMENTED** [review]
i think you erase the part  where the new balances are fullfilled D:

File: crates/pessimistic-proof/src/local_state.rs [hunk 1]
```diff
@@ -67,8 +67,12 @@ impl LocalNetworkState {
         }
 
         // TODO: benchmark if BTreeMap is the best choice in terms of SP1 cycles
-        let mut new_balances: BTreeMap<_, _> =
-            multi_batch_header.balances_proofs.iter().map(|(k, v)| (*k, v.0)).collect();
+        let mut new_balances = BTreeMap::new();
+        for (k, v) in &multi_batch_header.balances_proofs {
+            if new_balances.insert(*k, v.0).is_some() {
+                return Err(ProofError::DuplicateTokenBalanceProof);
```
Comment (Freyskeyd) [inline crates/pessimistic-proof/src/local_state.rs:73]: Could we add information about the duplication in order to log them and detect potential problem with a CDK?

Comment (wborgeaud) [reply crates/pessimistic-proof/src/local_state.rs]: Done

**Comment by Ciornenichii** [comment]
Ciornenichii stepan

**Pull request merged**